parquet2 implementation backed by parquet2 feature gate

houqp opened the pull request:
decouple core from arrow

# Description

WIP parquet2 implementation. The goal of this PR is to implement full read support leveraging parquet2. Write support is out of the scope and should be added as follow up PRs. Arrow2 integration is also out of the scope and should be added through follow up PR.

Currently all read tests are passing:

```
cargo test --no-default-features --features=arrow2,parquet2 
```

A quick benchmark shows more than 50% performance boost on checkpoint deserialization. Only tested with a very tiny checkpoint from the golden dataset, I would expect the performance gap would be bigger for larger real world tables.

Todo:

- [x] clean up duplicated code
- [x] support parsing map type
- [x] support parsing list type
- [x] benchmark

# Related Issue(s)

blocks #310 

# Documentation

<!---
Share links to useful documentation
--->

houqp commented:
Sounds good @ritchie46 , I will complete the parquet parsing support for map and list this weekend. But the branch I have here right now should be enough to unblock ploars integration.

ahmedriza commented:
There are currently some issues in `parquet2` with regard to nested types.  See https://github.com/jorgecarleitao/arrow2/issues/937 for example.

andrei-ionescu commented:
@houqp, any updates on this?

houqp commented:
@andrei-ionescu I have implemented all the data types other than map and nested list, so it's very close to be complete. However, my time is limited now, so progress will be slow. Anyone if welcome to collaborate on this branch to push this over the finish line :)

houqp commented:
alright, this branch is now feature complete ;) now it's time to catch up to latest delta-rs and arrow2/parquet2 releases.

houqp commented on rust/tests/dataframe_test.rs:
```diff
@@ -1,89 +1,83 @@
-extern crate deltalake;
 #[cfg(feature = "rust-dataframe-ext")]
-extern crate rust_dataframe;
+mod rust_dataframe {
```
no change in test, just moved code into its own mod to make conditional compilation annotation cleaner.

roeap replied:
don't think it matters much but `#![cfg(feature = "rust-dataframe-ext")]` might also work?

As a side note I was wondering if this still needs to / should be supported? It seems the repository for rust_dataframe has been archived, or did just move somewhere else?

houqp commented:
@wjones127 @roeap ready for review.

roeap commented on rust/src/action/parquet2_read/dictionary/binary.rs:
```diff
@@ -0,0 +1,49 @@
+use parquet2::encoding::get_length;
+use parquet2::error::Error;
+
+// FIXME: replace vec with reference
```
should this be fixed in this PR?

houqp replied:
good catch, totally forgot about this one.

roeap commented on rust/src/action/parquet2_read/map.rs:
```diff
@@ -0,0 +1,113 @@
+// use std::collections::HashMap;
```
remove?

roeap commented on rust/src/action/parquet2_read/mod.rs:
```diff
@@ -0,0 +1,864 @@
+            "metaData" => deserialize_metadata_column_page,
+            "protocol" => deserialize_protocol_column_page,
+            "commitInfo" => deserialize_commit_info_column_page,
+            "cdc" => deserialize_cdc_column_page,
```
❤️

roeap reviewed (commented):
Only had some very minor comments. While not an expert on parquet reading, I felt the tests should cover that quite nicely!

The one thing I was wondering is the `impl`'s for the actions, or the action.rs file in general. It seems we have alot of feature flags there now, and maybe it would be cleaner to have the parquet specific stuff also moved into its own mod. Maybe adopt something ile the trait based approach in the new parquet2 implementation?  

As this also seems like a step towards supporting `polars`, we will face the question of having good integration points to support various backends / engines or however one wnats to call it :).

great work!

houqp commented:
> maybe it would be cleaner to have the parquet specific stuff also moved into its own mod

Good idea, I have moved all those code into a `parquet_read` mod to keep action lean.

I chatted this with Andy and Jorge at the Data+AI summit a couple of months ago, the long term goal is to develop an Arrow trait that allows users to switch between arrow-rs and arrow2 in different projects including datafusion and delta-rs. This will also open up the possibility for a 3rd GPU based arrow implementation ;)

roeap reviewed (approved):
LGTM!
